### ISSUE TITLE
Opponents cards appear in wrong order, Cards created do not show up until the second one is created

### ISSUE THREAD

**Issue by robertsmieja** [Epix37/Hearthstone-Deck-Tracker#1406]
Using the latest pre-release version.

Screenshot: http://i.imgur.com/v7zefcD.jpg

1) Living Roots appears in the middle of the opponent's cards played.

2) Laughing Sister was created, then the Emerald Drake at the bottom. The Laughing Sister only showed up after the Emerald Drake was created.

**Comment by Epix37** [comment]
Happens randomly or is it consistent?

**Comment by Epix37** [comment]
Wait... they are sorted alphabetically.. what?

**Comment by robertsmieja** [comment]
Card ordering appears to be random? Happened for my first game only: http://i.imgur.com/5PziLeJ.jpg

**Comment by robertsmieja** [comment]
Generated card issue also doesn't appear to be consistent, unless it's Ysera only: http://i.imgur.com/S6Flkvg.jpg

Feel free to close this issue as something that might happen only the first time? I'll reopen if I can reproduce.

**Comment by robertsmieja** [comment]
Alphabetical ordering is occurring again: http://i.imgur.com/pP2i8tM.jpg

**Comment by Epix37** [comment]
Does it fix itself or is that consistent throughout the game?

**Comment by robertsmieja** [comment]
It's consistent for that game, but then the next two or so games seem to be fine. I can't seem to find a pattern as to when it occurs.

**Comment by robertsmieja** [comment]
To add, I am on Windows 10 64-bit.

I'm trying to trace through the code to see if I can figure out what's wrong, but I'm getting lost. Is the variable that contains an opponent's deck is in GameV2.cs? Is it the GameStats?

This is a bit far fetched, but Is it possible that the order is wrong because I don't have a deck list for my opponent? I think that could happen if the Deck type from the replay and the in-progress game is shared.

**Comment by Epix37** [comment]
The sorting happens [here](https://github.com/Epix37/Hearthstone-Deck-Tracker/blob/master/Hearthstone%20Deck%20Tracker/Windows/OverlayWindow.xaml.cs#L1297), which calls [this](https://github.com/Epix37/Hearthstone-Deck-Tracker/blob/master/Hearthstone%20Deck%20Tracker/Utility/Helper.cs#L210).

**Comment by Epix37** [comment]
And the opponent card list is [this one](https://github.com/Epix37/Hearthstone-Deck-Tracker/blob/master/Hearthstone%20Deck%20Tracker/Hearthstone/Player.cs#L137)

**Comment by robertsmieja** [comment]
Is the order for this List guranteed? https://github.com/Epix37/Hearthstone-Deck-Tracker/blob/master/Hearthstone%20Deck%20Tracker/Utility/Helper.cs#L219

If this is out of order it could be sorted by Type or Name first. From my mostly Java background this happens often. Type is Minion and Spell? 

I'll see if I can play some more games and see if that happens, although looking at this line it seems that type can be "Hero Power" so I'm not sure it would even show up? https://github.com/Epix37/Hearthstone-Deck-Tracker/blob/master/Hearthstone%20Deck%20Tracker/Hearthstone/Card.cs#L130

**Comment by Epix37** [comment]
I have not changed `Helper.SortCardCollection` in about year. And it used to work.
Type is Minion, Spell, etc. yes.

**Comment by robertsmieja** [comment]
Here's a screenshot where it's not in alphabetical order: http://i.imgur.com/DvVx9A9.jpg

It fixed itself next opponent's turn.

**Comment by Epix37** [comment]
The first one looks like "order of play" and and type. Weird stuff...

**Comment by robertsmieja** [comment]
The order of play isn't actually accurate though. And it was fine on my turn.

**Comment by Epix37** [comment]
I think I may have already accidentally stated the solution. Changing it back from `OpponentListView.Items` to `OpponentListView.ItemsSource` seems to solve the problem. Not sure why I even changed it in the first place. Need to do some more testing to see if this is actually it.

**Comment by Epix37** [comment]
Played a few games, seems to be fine. Give this version a shot: https://github.com/Epix37/HDT-Test/releases/tag/v0.10.18PRE1

**Comment by robertsmieja** [comment]
Thanks for fixing this so quickly. So the problem ended up being a place where sorting wasn't being done?

**Comment by Epix37** [comment]
The problem was: when I created the card lists that are supposed to be displayed, I created new card instances instead of loading them from the database. In theory that's totally fine becuase, when e.g. the name is then accessed, the cards loads the data automatically. *However*, that does not happen for the cost. So when sorting the cards by cost before sorting them by name (which is the order it needs to happen in), the cost would not be loaded yet and therefore just be 0.
